module/zfs/dmu_send.c

```diff
@@ -1537,7 +1541,8 @@ restore_object(struct restorearg *ra, objset_t *os, struct drr_object *drro)
 		}
 	}
 
-	tx = dmu_tx_create(os);
+	count = DN_BONUSLEN_TO_COUNT(drro->drr_bonuslen);
+	tx = dmu_tx_create_dn_count(os, count);
```

Comment:
OK so you can create an object with a specific slot count (just like you can create an object with a specific type, bonus length, block size, etc).  I think that you should specify the object parameters when you allocate it, with dmu_object_alloc/claim/reclaim, rather than having it "figure it out" from the tx.

If the tx also needs to know the slot count, that's fine.  Ideally it would be specified with dmu_tx_hold_*().